engine/kv_engine_cleaner.cpp

```diff
@@ -617,22 +605,177 @@ void Cleaner::AdjustThread(size_t advice_thread_num) {
   }
 }
 
+double Cleaner::SearchOutdatedCollections() {
+  size_t limited_fetch_num = 0;
+  std::unique_lock<SpinMutex> queue_lock(outdated_collections_.queue_mtx_);
+  int64_t before_queue_size =
+      static_cast<int64_t>(outdated_collections_.hashlists.size() +
+                           outdated_collections_.lists.size() +
+                           outdated_collections_.skiplists.size());
+  {
+    std::unique_lock<std::mutex> skiplist_lock(kv_engine_->skiplists_mu_);
+    auto iter = kv_engine_->expirable_skiplists_.begin();
+    while (!kv_engine_->expirable_skiplists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto outdated_skiplist = *iter;
+      iter = kv_engine_->expirable_skiplists_.erase(iter);
+      outdated_collections_.skiplists.emplace(outdated_skiplist);
+      limited_fetch_num++;
+    }
+  }
+
+  {
+    std::unique_lock<std::mutex> list_lock(kv_engine_->lists_mu_);
+    auto iter = kv_engine_->lists_.begin();
+    while (!kv_engine_->lists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto expired_list = *iter;
+      iter = kv_engine_->lists_.erase(iter);
+      outdated_collections_.lists.emplace(expired_list);
+      limited_fetch_num++;
+    }
+  }
+
+  {
+    std::unique_lock<std::mutex> hlist_lock(kv_engine_->hlists_mu_);
+    auto iter = kv_engine_->hash_lists_.begin();
+    while (!kv_engine_->hash_lists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto expired_hash_list = *iter;
+      iter = kv_engine_->hash_lists_.erase(iter);
+      outdated_collections_.hashlists.emplace(expired_hash_list);
+      limited_fetch_num++;
+    }
+  }
+
+  int64_t after_queue_size =
+      static_cast<int64_t>(outdated_collections_.hashlists.size() +
+                           outdated_collections_.lists.size() +
+                           outdated_collections_.skiplists.size());
+
+  double diff_size_ratio =
+      (after_queue_size - before_queue_size) / max_thread_num_;
+  if (diff_size_ratio > 0) {
+    outdated_collections_.increase_ratio =
+        outdated_collections_.increase_ratio == 0
+            ? diff_size_ratio
+            : diff_size_ratio / outdated_collections_.increase_ratio;
+  }
+  return outdated_collections_.increase_ratio;
+}
+
+void Cleaner::FetchOutdatedCollections(
+    PendingCleanRecords& pending_clean_records) {
+  Collection* outdated_collection = nullptr;
+  RecordType record_type;
+  auto min_snapshot_ts =
+      kv_engine_->version_controller_.GlobalOldestSnapshotTs();
```

Comment:
This `GlobalOldestSnapshotTs()` should be compared to in `SearchOutdatedCollections()`.
We should have clear definitions for different states of collections.
`Expirable`/`Persistent`: The collection is associated with an expire time or not.
Expired: The collection is manually deleted or current time has passed its expire time. Deleted collection is a subset of Expired collection as deletion is equivalent to expiring the collection with TTL=0.
`Outdated`: The collection has expired, and no global snapshot(user snapshot created by GetSnapshot/Iterator) is pinning the collection. Outdated collections are a subset of Expired collections.
`Destructible`: The collection is outdated, and no pointers are pointing to it. An outdated collection is removed from HashTable, std::set/std::vector/Collection::old_version so that there is no thread can access it except for the background cleaner threads. Cleaner threads must wait for all access tokens(local snapshots) has expired so that indeed no thread is visiting the Collection. Then they can do the actual destruction.

Reply:
`SearchOutdatedCollections()` only find the outdated collections (expired and deleted).  `FetchOutdatedCollections` will fetch a outdated collection which isn't held by any user's snapshot.